File: domains/org/src/Entities/Capability.cs
using Base.Entities;

namespace Org.Entities;

public class Capability : EntityBase
{
  public required int RoleId { get; set; }
  public Role Role { get; set; } = null!;

  public required int CapabilityTypeId { get; set; }
  public CapabilityType CapabilityType { get; set; } = null!;

  public DateTimeOffset ActiveFromDate { get; set; } = DateTimeOffset.UtcNow;
  public DateTimeOffset? RetiredAtDate { get; set; }

  public List<HomeAssignment> HomeAssignments { get; set; } = new();
  public List<Deployment> Deployments { get; set; } = new();
}

File: webapp/src/Client/Capability.ts
import { CapabilityType } from "./CapabilityTypes";
import { useCRUD, UseCRUDOptionalProps } from "./CRUD/CRUD";
import { Deployment } from "./Deployments";
import { HomeAssignment } from "./HomeAssignment";
import { RoleType } from "./RoleType";

export interface Capability {
  id: number;
  role_id: number;
  role: RoleType;
  capability_type_id: number;
  capability_type: CapabilityType;
  active_from_date: string;
  retired_at_date?: string;
  active_home_assignment?: HomeAssignment;
  active_deployment?: Deployment;
}

export interface MutateCapability {
  role_id: number;
  capability_type_id: number;
  active_from_date?: string;
  retired_at_date?: string;
}

export interface UseCapabilityProps extends UseCRUDOptionalProps {
  roleId?: number;
}

export function useCapability(props?: UseCapabilityProps) {
  const {
    reload,
    items,
    loading,
    error,
    createItem,
    retrieveItem,
    updateItem,
    deleteItem,
  } = useCRUD<Capability, MutateCapability>({
    path: props?.roleId
      ? `/api/org/roles/${props!.roleId}/capabilities`
      : "/api/org/capabilities",
    ...props,
  });
  return {
    items,
    loading,
    error,
    reload,
    createItem,
    retrieveItem,
    updateItem,
    deleteItem,
  };
}
